[WIP] Cohort course-wide and inline discussion topics UI.

muzaffaryousaf opened the pull request:
This adds the UI to the cohort management tab on LMS side. It allows the staff to cohort the course-wide as well as, inline discussions via friendly UI. 

The Sandbox is available at: http://cohort-discussions.m.sandbox.edx.org/

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+<div class="cohort-discussions-nav">
+    <form action="" method="post" id="cohort-coursewide-discussions-form" class="cohort-coursewide-discussions-form">
+        <h3 class="subsection-title"><%- gettext('Designate discussion topics as divided by cohort') %></h3>
```
@explorerleslie @marcotuts @muzaffaryousaf  I have a couple of questions/thoughts:
What behavior do we want after we click on the heading? I found it a bit disorienting that the section heading changed after I clicked "Assign cohorts to discussion topics" -- this doesn't happen with "Assign studnets to cohorts by uploading a CSV file". Do we want to make the behavior consistent? Keep heading the same after clicking and section expands? 
If yes, we should decide on wording for the heading. I think that "Assign cohorts to discussion topics" is not actually what we're doing here. Suggested heading: "Specify whether discussion topics are divided by cohort".

explorerleslie replied:
@catong I think that's just an implemented error in the wireframes. @muzaffaryousaf, please change the link text and the heading text once the link is clicked to this text: "Specify whether discussion topics are divided by cohort".

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+            <div class="form-fields">
+                <div class="form-field">
+                    <div class="coursewide-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Coursewide Discussion Topics') %></h3>
```
Add hyphen: "Course-wide Discussion Topics"

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+            <div class="form-fields">
+                <div class="form-field">
+                    <div class="inline-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Content Specific Discussion Topics') %></h3>
```
Add hyphen: "Content-Specific Discussion Topics"

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+                <div class="form-field">
+                    <div class="coursewide-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Coursewide Discussion Topics') %></h3>
+                        <p><%- gettext('You can optionally designate coursewide discussions topics to be cohorted.') %></p>
```
If it's in keeping with what we do in other areas, I would like to suggest making this a more active instruction: "Select the course-wide discussion topics that you want to divide by cohort." @explorerleslie @marcotuts @muzaffaryousaf

explorerleslie replied:
Looks good.

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+                <div class="form-field">
+                    <div class="inline-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Content Specific Discussion Topics') %></h3>
+                        <p><%- gettext('All other content-specific discussion topics, defined through inline discussion components, can also be designated as cohorted. You can cohort all discussion topics, or control specific areas below.') %></p>
```
Same comment as above, about making this text more active guidance.
Suggested text: "Select the content-specific discussion topics that you want to divide by cohort. To specify that all existing and future content-specific discussion topics are divided by cohort, select **Always Cohort Content-Specific Discussion Topics**.
@explorerleslie @marcotuts @muzaffaryousaf

explorerleslie replied:
Looks good.

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,49 @@
+                        <p><%- gettext('All other content-specific discussion topics, defined through inline discussion components, can also be designated as cohorted. You can cohort all discussion topics, or control specific areas below.') %></p>
+                        <div class="always_cohort_inline_discussions">
+                            <input type="checkbox" class="check-all-inline-discussions" <%- always_cohort_inline_discussions ? 'checked="checked"' : '' %>/>
+                            <span class="all-inline-discussions"><%= interpolate('Cohort all content specific discussion topics') %></span>
```
I think we discussed changing the label on this checkbox to make it more clear that it's not just an "All" checkbox? Do we want to call it "Always cohort content-specific discussion topics"?

explorerleslie replied:
Yes, I like this suggestion. @muzaffaryousaf please make the change.

catong commented on lms/templates/instructor/instructor_dashboard_2/cohorts.underscore:
```diff
@@ -47,5 +47,12 @@
+        <hr class="divider divider-lv1" />
+
+        <!-- Discussion Topics. -->
+        <a class="toggle-cohort-management-discussions" href="#cohort-discussions-management"><%- gettext('Assign cohorts to discussion topics') %></a>
```
See at line 3 of lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore about behavior on clicking heading. Suggested new heading might be "Specify whether discussion topics are divided by cohort". @explorerleslie @marcotuts your thoughts?

explorerleslie replied:
addressed in previous comment. This text should be changed to ""Specify whether discussion topics are divided by cohort".

catong commented:
After making selections and clicking Save, I do not see that discussions are marked with the blue "- Cohorted" label next to the discussion topic name, although I do see a message indicating "Changes Saved."

muzaffaryousaf commented:
@catong that's what we discussed in meeting that you would't be to see the changes until i rebase my branch from usman's branch. Will let you know when i'm done with that.

catong commented:
@muzaffaryousaf Oops, sorry -- thanks!

explorerleslie commented:
@muzaffaryousaf the interaction for the "Cohort all content specific discussion topics" checkbox is a little off. The behavior should be: 

-When the "Always cohort" checkbox is checked, all checkboxes below should be greyed out. None of the items below should change state (if a topic was checked prior to the "always cohort" checkbox being checked, it should remain checked, and the reverse). 
- When the "Always cohort" checkbox is unchecked, all checkboxes should become un-greyed out (available). Again, none of their state should change. 

@marcotuts please confirm.

marcotuts commented:
@explorerleslie - The disabled state of the checkboxes when the "Always cohort" toggle is selected is accurate as I understand it. You shouldn't be able to enable / disable these if they're being forced to be cohorted by the top setting.

explorerleslie commented:
![image](https://cloud.githubusercontent.com/assets/5643426/6445747/26a6f954-c0d4-11e4-9968-2bc6bf735ea3.png)

@muzaffaryousaf how do we end up with empty subcategory discussion topics? This is from http://cohort-discussions.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/instructor#view-cohort_management

explorerleslie commented:
@muzaffaryousaf please display course-wide topics in alphabetical order.

explorerleslie commented:
@muzaffaryousaf have you tested this with unicode characters in discussion topics? This has bitten us in the past. Please test with unicode and let me know the results. Thanks!

explorerleslie commented:
@muzaffaryousaf I'm done with this pass of reviews, and we're looking good. I left a bunch of comments. Please address comments and make saving work, and then re-ping me for a second review. Thanks!

muzaffaryousaf commented:
@explorerleslie i've tested with special characters, everything is working as expected, no weird behaviour.

muzaffaryousaf commented:
@catong i've addressed your comments. please re-review thanks.

muzaffaryousaf commented:
@explorerleslie looks like the empty subcategory discussion topics is bug in the existing code. please see in the screenshot. i guess we should fix this in separate ticket.
![empty discussion topic](https://cloud.githubusercontent.com/assets/2994509/6461072/898b3360-c1be-11e4-86ca-ecd20cac61e5.png)

muzaffaryousaf commented:
@explorerleslie i've addressed all of your comments and also implemented the save work. @catong @marcotuts can you please re-review the changes.  
Thanks.

muzaffaryousaf commented:
@explorerleslie @catong @marcotuts please continue review on PR https://github.com/edx/edx-platform/pull/7199 .I've rebased and opened this PR for saving work. 
Thanks.

catong commented:
@muzaffaryousaf I think my previous sugggested wording for UI text in the Content-Specific Discussion Topics section was less than ideal. After chatting with @explorerleslie and @marcotuts about the intended behavior I think this revised wording will cause less confusion for users. My apologies for the change. @explorerleslie @marcotuts Do you think this new wording is helpful?

Content-specific discussion topics are by default always divided by cohort. To make some content-specific discussion topics accessible to all learners regardless of cohort, clear **Always cohort content-specific discussion topics**, then clear the checkboxes for the individual discussion topics that you want to make unified.

Selecting **Always cohort content-specific discussion topics** overrides existing settings and makes all content-specific discussion topics divided by cohort.
